### PR TITLE
Link to wiki's solution page while working on or after completing cha…

### LINKED ISSUE #3406: After I complete a challenge, I can compare my solution with FCC's recommended solution
Ideally this would be a button like the one below (on the congratulations modal).

Edited by @bugron.
It would be great if we could have two code editors pop up side by side - one with the camper's solution and one with our recommended solution.

### LINKED ISSUE #8108: When opening wiki, open article for corresponding challenge
Agile user story:

- As a user, when I click the wiki button, the wiki should slide out with the article that corresponds to my current challenge, if one exists.

### PR DESCRIPTION
**Pull request by vcooley**
…llenge

#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of FreeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](https://github.com/FreeCodeCamp/FreeCodeCamp/wiki/git-rebase#squashing-multiple-commits-into-one) them into one commit).
- [x] All new and existing tests pass the command `npm run test-challenges`. Use `git commit --amend` to amend any fixes.
#### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Add new translation (feature adding new translations)
#### Checklist:

- [x] Tested changes locally.
- [x] Closes currently open issue (`Closes #XXXX`): Closes 
#### Description

Added functionality to the Wiki button in the navbar. Also added a button to the challenge complete modal.

If you are on a challenge that is of type HTML, JS, or Bonfire, clicking the Wiki button in the navbar or the new button in the challenge completion modal will take you to the appropriate challenge page that often has a solution.

Clicking one of these buttons while not on an HTML, JS, or Bonfire challenge will take you to the wiki homepage, as specific pages for these challenges and steps often do not exist.

Known issue:
There are some challenges (I've found 5 so far) that do not have a corresponding wiki page and thus bring up a 404 in the wiki pane.

closes #8108 
closes https://github.com/FreeCodeCamp/wiki-generator/issues/12

### CODE REVIEW

**Comment by raisedadead** [comment]
@vcooley Thanks for the PR, but IMHO this does not address, the needs of #3406 .

> It would be great if **we could have two code editors pop up side by side** - one with the camper's solution and one with our recommended solution.

This is what we are looking for, but it sure it does address #8108 

Would you mind in this case changing the closing tags, to reflect only #8108

**Comment by vcooley** [comment]
Done!

**Comment by raisedadead** [comment]
Yup, I do understand that, but nevertheless this is taken care by #8108 and #3406 is bit different from the current use case, hence we don't want to close #3406 with this PR actually.

Please feel free to continue to discussion in the issue thread. and thanks for contributing.

Will QA and let ya know on this PR.

File: client/main.js [hunk 1]
```diff
@@ -390,18 +390,40 @@ $(document).ready(function() {
...
+  }
+
   function showWiki() {
     if (!main.isWikiAsideLoad) {
       var lang = window.location.toString().match(/\/\w{2}\//);
       lang = (lang) ? lang[0] : '/en/';
       var wikiURL = '//freecodecamp.github.io/wiki' + lang;
+      var wikiPath = getWikiPath();
+      if (wikiPath) {
+        wikiURL = '//freecodecamp.github.io/wiki/en/' + wikiPath;
```
Comment (raisedadead) [inline client/main.js]: You are hard coding the language here. 
Should be `wikiURL = wikiURL + wikiPath`

Also I think we need a trailing `/` as well. Not sure, but I have seen in cases this has broken without it.

Comment (raisedadead) [inline client/main.js]: Yup nothing is rendered if you do not append the `/`

File: server/views/challenges/showBonfire.jade [hunk 2]
```diff
@@ -80,6 +80,7 @@ block content
                                   = phrase
                       else
                           a#next-challenge.btn.btn-lg.btn-primary.btn-block(href="/challenges/next-challenge?id="+id) Go to my next challenge (ctrl + enter)
+                      #show-solution.animated.fadeIn.btn.btn-lg.btn-primary.btn-block See the recommended solution
```
Comment (raisedadead) [inline server/views/challenges/showBonfire.jade]: Let's address this separately as discussed in the other issue for showing a recommended solution side by side.
You need to remove the buttons for now. Leave a comment on the other issue, if confirmed, make a separate PR

File: server/views/challenges/showHTML.jade [hunk 3]
```diff
@@ -70,6 +70,7 @@ block content
                                 #submit-challenge.animated.fadeIn.btn.btn-lg.btn-primary.btn-block Submit and go to my next challenge (ctrl + enter)
                             else
                                 a#next-challenge.btn.btn-lg.btn-primary.btn-block(href="/challenges/next-challenge?id="+id) Go to my next challenge (ctrl + enter)
+                            #show-solution.animated.fadeIn.btn.btn-lg.btn-primary.btn-block See the recommended solution
```
Comment (raisedadead) [inline server/views/challenges/showHTML.jade]: Let's address this separately as discussed in the other issue for showing a recommended solution side by side.
You need to remove the buttons for now. Leave a comment on the other issue, if confirmed, make a separate PR

File: server/views/challenges/showJS.jade [hunk 4]
```diff
@@ -75,6 +75,7 @@ block content
                             #submit-challenge.animated.fadeIn.btn.btn-lg.btn-primary.btn-block Submit and go to my next challenge (ctrl + enter)
                         else
                             a#next-challenge.animated.fadeIn.btn.btn-lg.btn-primary.btn-block(href="/challenges/next-challenge?id="+id) Go to my next challenge (ctrl + enter)
+                        #show-solution.animated.fadeIn.btn.btn-lg.btn-primary.btn-block See the recommended solution
```
Comment (raisedadead) [inline server/views/challenges/showJS.jade]: Let's address this separately as discussed in the other issue for showing a recommended solution side by side.
You need to remove the buttons for now. Leave a comment on the other issue, if confirmed, make a separate PR

**Comment by vcooley** [comment]
@raisedadead 

Thanks for the feedback, I'll get this fixed up this afternoon.

File: client/main.js (hunk 1 above)
Comment (vcooley) [inline client/main.js]: @raisedadead 
Hey I hardcoded the language because as far as I could tell, most challenge articles only have an english version. 

I don't think directing them to 404s is a very good idea, so at least for now directing them to the challenge page for their language seems like a bad choice. So, if a user's language is not english, should we just direct them to the wiki homepage?

What I was thinking in making it english is that we at least present some solution. This might benefit some bilingual people whose language is set to their native language.

Comment (raisedadead) [inline client/main.js]: Problem is this things would easily get overlooked. Hard coding right now is like introducing a bug.

Notice that `lang` is set from the URL and this logic should be used.
Additionally if there is no article then we should re-direct the page to the English version.

Comment (vcooley) [inline client/main.js]: So do you have any ideas on this? I've been doing reading and I see a couple potential solutions. There doesn't seem to be simple and reliable client side solution to check whether a page will send a 404 due to cross-origin policy.

1. (Hacky and fragile) Check the iframe's contents upon loading for a 404 message.

2. Send request to FCC's main server with the wiki path to verify that the wiki path is available in that language.

3. Change the wiki server so that it redirects to the english page on a 404.

Let me know if any of these directions sounds good or if I'm missing something obvious. Thanks again for your help.

Comment (raisedadead) [inline client/main.js]: Yeah I see this challenge in implementing this. The wiki is served from the gh-pages on GitHub.
So technically option 3 above is not possible (there is no wiki server per say).

That leaves us with option 1 and 2.

Option 2 is an overload because its an additional request and probably as you rightly said for every page.

While I do not like option 1 it could be done for now, else people are never going to discover that we have wiki articles (however few now) in the language they want to use FCC.

/cc @FreeCodeCamp/issue-moderators  Or should we drop the entire thing and let this redirect to english always as currently implemented.

**Comment by vcooley** [comment]
Sorry for the delay. I've been looking into this more and it seems that there is likely no way to read an iframe's contents. I think we'll need to implement a solution on the wiki side to get this feature up. Not sure if this is possible in the current state of the wiki but I'm going to start looking at some of the wiki code to figure out if we can redirect on a 404.

**Comment by raisedadead** [comment]
Thanks for this update @vcooley much appreciated. In that case just remove the buttons as I mentioned, move the hard coding and add it on the top.

For now lets continue the way we have this and re-direct to English always.

Also raise a separate issue to track the problem.

File: client/main.js [hunk 5]
```diff
@@ -390,18 +390,47 @@ $(document).ready(function() {
...
   function showWiki() {
     if (!main.isWikiAsideLoad) {
+      var wikiURL = '//freecodecamp.github.io/wiki/en/';
       var lang = window.location.toString().match(/\/\w{2}\//);
       lang = (lang) ? lang[0] : '/en/';
-      var wikiURL = '//freecodecamp.github.io/wiki' + lang;
+      var wikiPath = getWikiPath();
+      if (wikiPath) {
+        wikiURL = wikiURL + wikiPath + '/';
+      }
```
Comment (raisedadead) [inline client/main.js]:
```
/home/travis/build/FreeCodeCamp/FreeCodeCamp/client/main.js
  430:12  error  Closing curly brace does not appear on the same line as the subsequent block  brace-style
```

Change this to :
```
} else if
```

**Comment by raisedadead** [comment]
@vcooley when you amend the commit could you use a smaller commit message it is recommended to have one lesser than 50 chars.

**Comment by vcooley** [comment]
@raisedadead 
Thank you! I could not figure out what to do to fix that eslint error. And my editor was not displaying the message so I was pretty confused. Everything should be good to go. I'll go ahead and raise an issue on the hard coding.

**Comment by raisedadead** [comment]
@vcooley and @Rafase282 this doesnt seem to work any more. Can anyone of you confirm unless i am doing something wrong while testing

**Comment by vcooley** [comment]
Where did the wiki open to? The wiki home page or a 404?

If it opened to the home page, make sure you have navigated to a challenge such as https://www.freecodecamp.com/challenges/pig-latin

I'll take a look at it in a few hours if that doesn't work for you.

**Comment by raisedadead** [comment]
It defaults to the Home page.

image

**Comment by vcooley** [comment]
Ok, thanks for letting me know. I'll look into it.

**Comment by vcooley** [comment]
It's working for me.

image

**Comment by Rafase282** [comment]
@raisedadead @vcooley maybe it had something to do with the wiki links being camel case before and now lowercase?

**Comment by vcooley** [comment]
I don't think so. It uses 'window.common.dashedName' which I believe should
be always lowercase. And I'd expect it to pull up a 404 page if there was a
discrepancy there. Maybe there's a problem with 'window.common' not being
initialized properly? I know it isn't declared on the homepage.

**Comment by raisedadead** [comment]
I can't seem to get this  working. 
requesting other's to test if possible. 
cc/ @FreeCodeCamp/issue-moderators

**Comment by Rafase282** [comment]
@vcooley @raisedadead how do I go about testing this?

**Comment by raisedadead** [comment]
@Rafase282 :

1. Pull down the PR.
2. seed the database (optional, but recommended) : `node seed`
3. Visit a challenge for ex. Pig Latin that has a wiki solution
4. Click on the wiki tab, you should see the wiki solution for that challenge instead of the home page.

**Comment by Rafase282** [comment]
Seems like I'm missing the `.env` file when trying node seed.

**Comment by Rafase282** [comment]
# LGTM +1

Works for all challenges not just algorithms.

Imgur

Imgur

File: client/main.js [hunk 6]
```diff
@@ -390,18 +390,46 @@ $(document).ready(function() {
   }
 
   $('#nav-wiki-btn').on('click', function(event) {
-      if (!(event.ctrlKey || event.metaKey)) {
-          toggleWiki();
-      }
+    if (!(event.ctrlKey || event.metaKey)) {
+      toggleWiki();
+    }
```
Comment (BerkeleyTrue) [inline client/main.js]: Thanks for fixing the indentation 👍

**Comment by BerkeleyTrue** [comment]
LGTM 👍

**Comment by BerkeleyTrue** [comment]
@vcooley Thanks so much for this!! ✨

**Comment by raisedadead** [comment]
@vcooley thanks.

**Comment by vcooley** [comment]
@raisedadead
Did you ever get it working?
